File: sources/iOS/ios-communications/Tests/iOSCommunicationsTests/SkinTemperatureDataTest.swift
///  Copyright © 2023 Polar. All rights reserved.

import XCTest
#if canImport(PolarBleSdkShared)
import PolarBleSdkShared
#endif
@testable import iOSCommunications

final class SkinTemperatureDataTest: XCTestCase {

    func testUncompressedSkinTemperatureData() throws {
        let previousTimeStamp: UInt64 = 120

        let temperatureDataFrameHeader = Data([
            0x07, 0x40, 0xAE, 0x21, 0xAE, 0x31, 0xB2, 0xEE, 0x0A, 0x00
        ])

        let temperatureDataFrameContent = Data([
            0xF6, 0x28, 0xC0, 0x41
        ])

        let factor:Float = 1.0
        let dataFrame = try PmdDataFrame(
            data: temperatureDataFrameHeader + temperatureDataFrameContent,
            { _,_ in previousTimeStamp },
            { _ in factor },
            { _ in 0 })

        let temperatureData = try SkinTemperatureData.parseDataFromDataFrame(frame: dataFrame)
        
        XCTAssertEqual(1,temperatureData.samples.count)
        XCTAssertEqual(24.0200005, temperatureData.samples[0].skinTemperature)
        XCTAssertEqual(787762911281000000, temperatureData.samples[0].timeStamp)
    }
    
    func testCompressedSkinTemperatureData() throws {
        let previousTimeStamp: UInt64 = 120
        
        
        let skinTemperatureDataFrameHeader = Data([
            0x07, 0x40, 0xAE, 0x21, 0xAE, 0x31, 0xB2, 0xEE, 0x0A, 0x80
        ])
        
        let skinTemperatureDataFrameContent = Data([
            0xEC, 0x51, 0xDC, 0x41, 0x03, 0x02, 0x00
        ])

        let factor:Float = 1.0
        let dataFrame = try PmdDataFrame(
            data: skinTemperatureDataFrameHeader + skinTemperatureDataFrameContent,
            { _,_ in previousTimeStamp },
            { _ in factor },
            { _ in 0 })

        let temperatureData = try SkinTemperatureData.parseDataFromDataFrame(frame: dataFrame)
        
        XCTAssertEqual(3,temperatureData.samples.count)
        XCTAssertEqual(27.54, temperatureData.samples[0].skinTemperature)
        XCTAssertEqual(27.54, temperatureData.samples[1].skinTemperature)
        XCTAssertEqual(27.54, temperatureData.samples[2].skinTemperature)
    }

    func testSkinTemperatureRawType0ParserUsesSharedKmpWhenLinked() throws {
        #if canImport(PolarBleSdkShared)
        let dataFrameHex = "0740ae21ae31b2ee0a00f628c041"
        let sharedRows = try XCTUnwrap(PolarIosSharedBridge.shared.skinTemperatureRawType0Samples(dataFrameHex: dataFrameHex, previousTimeStamp: 787762910281000000, factor: 1.0, sampleRate: 4))
        XCTAssertFalse(sharedRows.isEmpty)

        let dataFrame = try PmdDataFrame(
            data: Data(hexString: dataFrameHex),
            { _, _ in 787762910281000000 },
            { _ in 1.0 },
            { _ in 4 })
        let skinTemperatureData = try SkinTemperatureData.parseDataFromDataFrame(frame: dataFrame)
        let sharedSamples = try sharedRows.split(separator: "|").map { row -> (UInt64, Float) in
            let fields = row.split(separator: ",")
            return (
                try XCTUnwrap(UInt64(fields[0])),
                try XCTUnwrap(Float(fields[1]))
            )
        }

        XCTAssertEqual(sharedSamples.count, skinTemperatureData.samples.count)
        for (index, sharedSample) in sharedSamples.enumerated() {
            XCTAssertEqual(sharedSample.0, skinTemperatureData.samples[index].timeStamp)
            XCTAssertEqual(sharedSample.1, skinTemperatureData.samples[index].skinTemperature, accuracy: 0.00001)
            XCTAssertFalse(skinTemperatureData.samples[index].isTimestampEstimated)
        }
        #else
        throw XCTSkip("PolarBleSdkShared is not linked in this build")
        #endif
    }

    func testSkinTemperatureGoldenVectorsMatchIOSCommunicationsBehavior() throws {
        let vectors = try loadSkinTemperatureGoldenVectors()
        XCTAssertFalse(vectors.isEmpty, "Expected skin-temperature golden vectors")

        for vector in vectors {
            let id = vector["id"] as? String ?? "unknown-vector"
            if let platforms = vector["platforms"] as? [String: Any],
               let supported = platforms["ios"] as? Bool,
               !supported {
                continue
            }
            let input = try XCTUnwrap(vector["input"] as? [String: Any], id)
            let expected = try XCTUnwrap(vector["expected"] as? [String: Any], id)
            let dataFrame = try PmdDataFrame(
                data: Data(hexString: try XCTUnwrap(input["dataFrameHex"] as? String, id)),
                { _, _ in UInt64(truncating: input["previousTimeStamp"] as? NSNumber ?? 0) },
                { _ in Float(truncating: input["factor"] as? NSNumber ?? 1.0) },
                { _ in UInt(truncating: input["sampleRate"] as? NSNumber ?? 0) })

            if let parseError = expected["parseError"] as? String {
                XCTAssertThrowsError(try SkinTemperatureData.parseDataFromDataFrame(frame: dataFrame), id) { error in
                    switch parseError {
                    case "unsupportedFrame":
                        guard case BleGattException.gattDataError = error else {
                            return XCTFail("Expected gattDataError for \(id), got \(error)")
                        }
                    case "malformedFrame":
                        guard case BleGattException.gattDataError = error else {
                            return XCTFail("Expected gattDataError for \(id), got \(error)")
                        }
                    default:
                        XCTFail("Unsupported parse error expectation in \(id): \(parseError)")
                    }
                }
                XCTAssertEqual(try XCTUnwrap(expected["timeStamp"] as? NSNumber, id).uint64Value, dataFrame.timeStamp, id)
                continue
            }

            let skinTemperatureData = try SkinTemperatureData.parseDataFromDataFrame(frame: dataFrame)

            let samples = try XCTUnwrap(expected["samples"] as? [[String: Any]], id)
            XCTAssertEqual(samples.count, skinTemperatureData.samples.count, id)
            let platformExpectations = vector["platformExpectations"] as? [String: Any]
            let iosExpectations = platformExpectations?["ios"] as? [String: Any]
            let iosSamples = iosExpectations?["samples"] as? [[String: Any]]
            for (index, expectedSample) in samples.enumerated() {
                let actualSample = skinTemperatureData.samples[index]
                XCTAssertEqual(try XCTUnwrap(expectedSample["timeStamp"] as? NSNumber, id).uint64Value, actualSample.timeStamp, id)
                XCTAssertEqual(try XCTUnwrap(expectedSample["skinTemperature"] as? NSNumber, id).floatValue, actualSample.skinTemperature, accuracy: 0.00001, id)
                if let expectedEstimated = iosSamples?[index]["isTimestampEstimated"] as? Bool {
                    XCTAssertEqual(expectedEstimated, actualSample.isTimestampEstimated, id)
                }
            }
        }
    }

    func testSkinTemperatureGoldenVectorsFollowNeutralKmpShape() throws {
        for vector in try loadSkinTemperatureGoldenVectors() {
            let id = try XCTUnwrap(vector["id"] as? String)
            XCTAssertNotNil(vector["area"], id)
            XCTAssertNotNil(vector["case"], id)
            XCTAssertNotNil(vector["source"], id)
            XCTAssertNotNil(vector["input"], id)
            XCTAssertNotNil(vector["expected"], id)
            let platforms = try XCTUnwrap(vector["platforms"] as? [String: Any], id)
            XCTAssertNotNil(platforms["android"], id)
            XCTAssertNotNil(platforms["ios"], id)
            XCTAssertNotNil(platforms["common"], id)
        }
    }

    func testSkinTemperatureReadinessManifestIsPinnedBeforeParserMigration() throws {
        let manifest = try loadSkinTemperatureReadinessManifest()
        let id = try XCTUnwrap(manifest["id"] as? String)
        let input = try XCTUnwrap(manifest["input"] as? [String: Any], id)
        let expected = try XCTUnwrap(manifest["expected"] as? [String: Any], id)
        let consumerTests = try XCTUnwrap(manifest["consumerTests"] as? [String: Any], id)
        let requiredFamilies = try XCTUnwrap(input["requiredBehaviorFamilies"] as? [String], id)
        let coveredFamilies = try XCTUnwrap(expected["coveredBehaviorFamilies"] as? [String], id)
        let policyVectorPaths = try XCTUnwrap(input["policyVectorPaths"] as? [String], id)

        XCTAssertEqual("skin-temperature-readiness", id)
        XCTAssertEqual("skinTemperatureReadiness", input["kind"] as? String, id)
        XCTAssertEqual(SKIN_TEMPERATURE_READINESS_POLICY_VECTOR_PATHS, policyVectorPaths, id)
        XCTAssertEqual(SKIN_TEMPERATURE_READINESS_FAMILIES, requiredFamilies, id)
        XCTAssertEqual(SKIN_TEMPERATURE_READINESS_FAMILIES, coveredFamilies, id)
        XCTAssertEqual(["com.polar.androidcommunications.api.ble.model.gatt.client.pmd.model.SkinTemperatureDataTest"], consumerTests["android"] as? [String], id)
        XCTAssertEqual(["SkinTemperatureDataTest"], consumerTests["ios"] as? [String], id)
        XCTAssertEqual(["com.polar.sharedtest.SkinTemperatureParserCommonPolicyTest"], consumerTests["commonPrototype"] as? [String], id)
    }

    private func loadSkinTemperatureGoldenVectors() throws -> [[String: Any]] {
        let vectorDirectory = try GoldenVectorTestData.repositoryRoot()
            .appendingPathComponent("testdata/golden-vectors/protocol/sensors")
        return try FileManager.default
            .contentsOfDirectory(at: vectorDirectory, includingPropertiesForKeys: nil)
            .filter { $0.pathExtension == "json" && $0.lastPathComponent.hasPrefix("skin-temperature-") }
            .sorted { $0.lastPathComponent < $1.lastPathComponent }
            .map { file in
                let data = try Data(contentsOf: file)
                return try XCTUnwrap(JSONSerialization.jsonObject(with: data) as? [String: Any], file.path)
            }
            .filter { vector in
                guard let input = vector["input"] as? [String: Any] else {
                    return true
                }
                return input["kind"] as? String != "skinTemperatureReadiness"
            }
    }

    private func loadSkinTemperatureReadinessManifest() throws -> [String: Any] {
        let vectorFile = try GoldenVectorTestData.repositoryRoot()
            .appendingPathComponent("testdata/golden-vectors/protocol/sensors/skin-temperature-readiness.json")
        let data = try Data(contentsOf: vectorFile)
        return try XCTUnwrap(JSONSerialization.jsonObject(with: data) as? [String: Any], vectorFile.path)
    }


    private let SKIN_TEMPERATURE_READINESS_POLICY_VECTOR_PATHS = [
        "protocol/sensors/skin-temperature-raw-type0-estimated-sample-rate.json",
        "protocol/sensors/skin-temperature-raw-type0-single-sample.json",
        "protocol/sensors/skin-temperature-raw-type0-truncated-sample-android-error.json",
        "protocol/sensors/skin-temperature-raw-type0-truncated-sample-ios-empty.json",
        "protocol/sensors/skin-temperature-raw-type1-unsupported.json"
    ]

    private let SKIN_TEMPERATURE_READINESS_FAMILIES = [
        "raw-type0-ieee754-skin-temperature-parsing",
        "sample-rate-timestamp-estimation-policy",
        "unsupported-raw-frame-policy",
        "truncated-raw-sample-policy",
        "ios-empty-malformed-payload-deferral",
        "platform-skin-temperature-vector-reference-gate",
        "compile-verification-gate"
    ]
}

private extension Data {
    init(hexString: String) throws {
        guard hexString.count.isMultiple(of: 2) else {
            throw NSError(domain: "SkinTemperatureDataTest", code: 2, userInfo: [NSLocalizedDescriptionKey: "Hex string must have an even length"])
        }
        var bytes: [UInt8] = []
        var index = hexString.startIndex
        while index < hexString.endIndex {
            let nextIndex = hexString.index(index, offsetBy: 2)
            let byteString = String(hexString[index..<nextIndex])
            guard let byte = UInt8(byteString, radix: 16) else {
                throw NSError(domain: "SkinTemperatureDataTest", code: 3, userInfo: [NSLocalizedDescriptionKey: "Invalid hex byte \(byteString)"])
            }
            bytes.append(byte)
            index = nextIndex
        }
        self.init(bytes)
    }
}

File: sources/Android/android-communications/library/src/test/java/com/polar/androidcommunications/api/ble/model/gatt/client/pmd/model/SkinTemperatureDataTest.kt
package com.polar.androidcommunications.api.ble.model.gatt.client.pmd.model

import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.PmdDataFrame
import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.PmdMeasurementType
import org.junit.Assert
import org.junit.Test
import java.io.File
import java.io.FileReader

class SkinTemperatureDataTest {

    @Test
    fun `dataFromRawType0_parse-data_with-not-compresed-data`() {

        // 10 first values form the header part of the data frame. The rest is temperature data.
        // 0x07(decimal 7) = measurement type for skin temp
        // 0x40...0x0A = timestamp
        // 0x00(decimal 0) = uncompressed, raw data. Results 0 (false) for check compressed mask (0x80, decimal 128)
        // 0x00...0x41 the uncompressed skin temperature value
        val temperatureDataFrame = byteArrayOf(
            0x07.toByte(), 0x40.toByte(), 0xAE.toByte(), 0x21.toByte(), 0xAE.toByte(), 0x31.toByte(),
            0xB2.toByte(), 0xEE.toByte(), 0x0A.toByte(), 0x00.toByte(), 0x00.toByte(), 0x60.toByte(),
            0xEE.toByte(), 0x41.toByte()
        )
        val timeStamp = 787762911281000000uL
        val previousTimeStamp = 787762910281000000uL

        val factor = 1.0f
        val dataFrame = PmdDataFrame(
            data = temperatureDataFrame,
            getPreviousTimeStamp = { pmdMeasurementType: PmdMeasurementType, pmdDataFrameType: PmdDataFrame.PmdDataFrameType -> previousTimeStamp },
            getFactor = { factor }
        ) { 0 }

        // Act
        val temperatureData = SkinTemperatureData.parseDataFromDataFrame(dataFrame)

        // Assert
        Assert.assertEquals(1, temperatureData.skinTemperatureSamples.size)
        Assert.assertEquals(29.796875f, temperatureData.skinTemperatureSamples[0].skinTemperature)
        Assert.assertEquals(timeStamp, temperatureData.skinTemperatureSamples[0].timeStamp)
    }

    @Test
    fun `dataFromCompressedType0_parse-data_with-compressed-data`() {

        // 10 first values form the header part of the data frame. The rest is temperature data.
        // 0x07(decimal 7) = measurement type for skin temp
        // 0x40...0x0A = timestamp
        // 0x80(decimal 128) = compressed data. Results 1 (true for check compressed mask (0x80, decimal 128)
        // 0xEC...0x00 the delta compressed skin temperature values
        val temperatureDataFrame = byteArrayOf(
            0x07.toByte(), 0x40.toByte(), 0xAE.toByte(), 0x21.toByte(), 0xAE.toByte(), 0x31.toByte(),
            0xB2.toByte(), 0xEE.toByte(), 0x0A.toByte(), 0x80.toByte(), 0xEC.toByte(), 0x51.toByte(),
            0xDC.toByte(), 0x41.toByte(), 0x03.toByte(), 0x02.toByte(), 0x00.toByte()
        )
        val dataFrame = PmdDataFrame(
            data = temperatureDataFrame,
            getPreviousTimeStamp = {  pmdMeasurementType: PmdMeasurementType, pmdDataFrameType: PmdDataFrame.PmdDataFrameType -> 1000uL },
            getFactor = { 1.0f }
        ) { 13 }

        // Act
        val temperatureData = SkinTemperatureData.parseDataFromDataFrame(dataFrame)

        // Assert
        Assert.assertEquals(3, temperatureData.skinTemperatureSamples.size)
        Assert.assertEquals(27.54f, temperatureData.skinTemperatureSamples[0].skinTemperature)
        Assert.assertEquals(27.54f, temperatureData.skinTemperatureSamples[1].skinTemperature)
        Assert.assertEquals(27.54f, temperatureData.skinTemperatureSamples[2].skinTemperature)
    }

    @Test
    fun skinTemperatureGoldenVectors_matchAndroidBehavior() {
        val vectors = loadSkinTemperatureVectors()
        Assert.assertTrue("Expected skin-temperature golden vectors", vectors.isNotEmpty())

        vectors.forEach { vector ->
            val caseId = vector.get("id").asString
            if (vector.getAsJsonObject("platforms")?.get("android")?.asBoolean == false) {
                return@forEach
            }
            val input = vector.getAsJsonObject("input")
            val expected = vector.getAsJsonObject("expected")
            val frame = PmdDataFrame(
                data = input.get("dataFrameHex").asString.hexToByteArray(),
                getPreviousTimeStamp = { _, _ -> input.get("previousTimeStamp").asLong.toULong() },
                getFactor = { input.get("factor").asFloat },
                getSampleRate = { input.get("sampleRate").asInt }
            )

            if (expected.has("parseError")) {
                assertParseError(caseId, expected.get("parseError").asString, frame)
                Assert.assertEquals(caseId, expected.get("timeStamp").asLong.toULong(), frame.timeStamp)
                return@forEach
            }

            val skinTemperatureData = SkinTemperatureData.parseDataFromDataFrame(frame)

            Assert.assertEquals(caseId, expected.get("timeStamp").asLong.toULong(), frame.timeStamp)
            assertSkinTemperatureSamples(caseId, expected.getAsJsonArray("samples"), skinTemperatureData.skinTemperatureSamples)
        }
    }

    private fun assertParseError(caseId: String, expectedError: String, frame: PmdDataFrame) {
        when (expectedError) {
            "unsupportedFrame" -> Assert.assertThrows(caseId, Exception::class.java) {
                SkinTemperatureData.parseDataFromDataFrame(frame)
            }
            "malformedFrame" -> Assert.assertThrows(caseId, Exception::class.java) {
                SkinTemperatureData.parseDataFromDataFrame(frame)
            }
            else -> Assert.fail("$caseId has unsupported parse error expectation $expectedError")
        }
    }

    @Test
    fun `skin temperature golden vectors follow neutral KMP vector shape`() {
        loadSkinTemperatureVectors().forEach { vector ->
            val id = vector.get("id").asString
            Assert.assertTrue(id, vector.has("area"))
            Assert.assertTrue(id, vector.has("case"))
            Assert.assertTrue(id, vector.has("source"))
            Assert.assertTrue(id, vector.has("input"))
            Assert.assertTrue(id, vector.has("expected"))
            Assert.assertTrue(id, vector.has("platforms"))
            val platforms = vector.getAsJsonObject("platforms")
            Assert.assertTrue(id, platforms.has("android"))
            Assert.assertTrue(id, platforms.has("ios"))
            Assert.assertTrue(id, platforms.has("common"))
        }
    }

    @Test
    fun `skin temperature readiness manifest is pinned before parser migration`() {
        val manifest = loadSkinTemperatureReadinessManifest()
        val id = manifest.get("id").asString
        val input = manifest.getAsJsonObject("input")
        val expected = manifest.getAsJsonObject("expected")
        val consumerTests = manifest.getAsJsonObject("consumerTests")
        val requiredFamilies = input.getAsJsonArray("requiredBehaviorFamilies").map { it.asString }
        val coveredFamilies = expected.getAsJsonArray("coveredBehaviorFamilies").map { it.asString }
        val policyVectorPaths = input.getAsJsonArray("policyVectorPaths").map { it.asString }

        Assert.assertEquals("skin-temperature-readiness", id)
        Assert.assertEquals("skinTemperatureReadiness", input.get("kind").asString)
        Assert.assertEquals(id, SKIN_TEMPERATURE_READINESS_POLICY_VECTOR_PATHS, policyVectorPaths)
        Assert.assertEquals(id, SKIN_TEMPERATURE_READINESS_FAMILIES, requiredFamilies)
        Assert.assertEquals(id, SKIN_TEMPERATURE_READINESS_FAMILIES, coveredFamilies)
        Assert.assertEquals(id, listOf("com.polar.androidcommunications.api.ble.model.gatt.client.pmd.model.SkinTemperatureDataTest"), consumerTests.getAsJsonArray("android").map { it.asString })
        Assert.assertEquals(id, listOf("SkinTemperatureDataTest"), consumerTests.getAsJsonArray("ios").map { it.asString })
        Assert.assertEquals(id, listOf("com.polar.sharedtest.SkinTemperatureParserCommonPolicyTest"), consumerTests.getAsJsonArray("commonPrototype").map { it.asString })
    }

    private fun assertSkinTemperatureSamples(
        caseId: String,
        expectedSamples: JsonArray,
        actualSamples: List<SkinTemperatureData.SkinTemperatureSample>
    ) {
        Assert.assertEquals(caseId, expectedSamples.size(), actualSamples.size)
        expectedSamples.forEachIndexed { index, expectedSample ->
            val sample = expectedSample.asJsonObject
            val actualSample = actualSamples[index]
            Assert.assertEquals(caseId, sample.get("timeStamp").asLong.toULong(), actualSample.timeStamp)
            Assert.assertEquals(caseId, sample.get("skinTemperature").asFloat, actualSample.skinTemperature)
        }
    }

    private fun loadSkinTemperatureVectors(): List<JsonObject> {
        val vectorDirectory = findRepositoryRoot()
            .resolve("testdata/golden-vectors/protocol/sensors")
        return vectorDirectory
            .listFiles { file -> file.isFile && file.extension == "json" && file.name.startsWith("skin-temperature-") }
            .orEmpty()
            .sortedBy { it.name }
            .map { file ->
                FileReader(file).use { reader ->
                    JsonParser.parseReader(reader).asJsonObject
                }
            }
            .filter { vector -> vector.getAsJsonObject("input").get("kind")?.asString != "skinTemperatureReadiness" }
    }

    private fun loadSkinTemperatureReadinessManifest(): JsonObject {
        val vectorFile = findRepositoryRoot()
            .resolve("testdata/golden-vectors/protocol/sensors/skin-temperature-readiness.json")
        FileReader(vectorFile).use { reader ->
            return JsonParser.parseReader(reader).asJsonObject
        }
    }

    private fun findRepositoryRoot(): File {
        val userDirectory = System.getProperty("user.dir") ?: error("user.dir is not set")
        var directory = File(userDirectory).absoluteFile
        while (true) {
            if (directory.resolve("testdata/golden-vectors").isDirectory) {
                return directory
            }
            directory = directory.parentFile ?: error("Could not find repository root from $userDirectory")
        }
    }

    private fun String.hexToByteArray(): ByteArray {
        require(length % 2 == 0) { "Hex string must have an even length" }
        return chunked(2)
            .map { it.toInt(16).toByte() }
            .toByteArray()
    }

    private companion object {
        val SKIN_TEMPERATURE_READINESS_POLICY_VECTOR_PATHS = listOf(
            "protocol/sensors/skin-temperature-raw-type0-estimated-sample-rate.json",
            "protocol/sensors/skin-temperature-raw-type0-single-sample.json",
            "protocol/sensors/skin-temperature-raw-type0-truncated-sample-android-error.json",
            "protocol/sensors/skin-temperature-raw-type0-truncated-sample-ios-empty.json",
            "protocol/sensors/skin-temperature-raw-type1-unsupported.json"
        )

        val SKIN_TEMPERATURE_READINESS_FAMILIES = listOf(
            "raw-type0-ieee754-skin-temperature-parsing",
            "sample-rate-timestamp-estimation-policy",
            "unsupported-raw-frame-policy",
            "truncated-raw-sample-policy",
            "ios-empty-malformed-payload-deferral",
            "platform-skin-temperature-vector-reference-gate",
            "compile-verification-gate"
        )
    }
}
